CMake: untie fvtest from GC

samolisov opened the pull request:
CMake: build fvtest without omrgc

The following problems related to build fvtest without omrgc (`OMR_FVTEST=ON OMR_GC=OFF`) were solved:

 * `${OMR_GC_LIB}` was removed from all fvtest targets (except the
     omrgctest one): omrcore is linked against omrgc if omrgc is enabled.

 * omrcore is linked against the `${OMR_CORE_GLUE_TARGET}` and
     `${OMR_RAS_GLUE_TARGET}` libraries.

 * omrcore is linked against `${OMR_TRACE_LIB}` and `${OMR_HOOK_LIB}`, these libraries are removed 
    from targets where they explicitly appear.

Closes: #2481
Signed-off-by: Pavel Samolysov <samolisov@gmail.com>

charliegracie commented:
@genie-omr build all

charliegracie commented:
@dnakamura and @rwy0717 this looks good to me.  Further improvements around RAS/RASGlue can be made as I go through and properly wrap the RAS code with its build flag.

charliegracie commented on omr/CMakeLists.txt:
```diff
@@ -71,14 +84,6 @@ if(OMR_GC)
-	target_include_directories(omrcore
-		PRIVATE
-			$<TARGET_PROPERTY:omrgc,INTERFACE_INCLUDE_DIRECTORIES>
-			$<TARGET_PROPERTY:${OMR_GC_GLUE_TARGET},INTERFACE_INCLUDE_DIRECTORIES>
```
GC glue should only be required by core if the GC is enabled right?

rwy0717 replied:
Yep.

samolisov replied:
@charliegracie @rwy0717 GC glue is exported by omrgc itself:
`gc/CMakeLists.txt`:

```cmake
target_include_directories(omrgc
	PUBLIC
		.
		${CMAKE_CURRENT_BINARY_DIR}  #Ideally this should be private, but the glue pulls this in
		base
		base/segregated
		base/standard
		include
		startup
		stats
		structs
		verbose
		verbose/handler_standard
	INTERFACE
		$<TARGET_PROPERTY:${OMR_GC_GLUE_TARGET},INTERFACE_INCLUDE_DIRECTORIES>
)
```

so, if omrcore links to omrgc (if `OMR_GC` is enabled), omrgc's `target_include_directories` will be added to the link interface of the omrcore target. I believe these lines may be removed without any consequences.

rwy0717 commented on example/CMakeLists.txt:
```diff
@@ -30,23 +30,26 @@ target_include_directories(omr_example_base
-# add_test(example example)
+# The current implementation of the example depends on the ${OMR_GC_LIB} library
+# and is not able to be compiled without a GC implementation.
+if(OMR_GC)
```
OK, that makes sense

samolisov replied:
OK, I've changed the `example` executable to `gcexample` one. So, if we wish to add another example, for jitbuilder, we should add a new executable to `CMakeLists.txt`:

```cmake
if(OMR_GC)
	add_executable(gcexample
		gcmain.cpp
	)

     ...
endif(OMR_GC)
...
if(OMR_JITBUILDER)
	add_executable(jitexample
		jitmain.cpp
	)

     ...
endif(OMR_JITBUILDER)
...
```

(I see the makefile now includes all .cpp files from the example folder to build the `omrgcexample` module, so the idea is also implemented for autoconf but when new examples are added, the makefile should be modified).

I'm not sure should we maintain the `example` test here or the test must be moved to `fvtest` where the other tests are?

samolisov commented:
@rwy0717 You wrote:

> This PR asserts that all FV tests are generic. Is this actually valid? Maybe we should consider be breaking the GC FV tests into two test suites, as a part of this PR?

I'm not sure I understand what a generic test means. My goal was to be able to build and test OMR components without the heaviest one - GC (for example, build and run fvtest against the compiler). I think if OMR is positioned as a set of cross-platform components for building reliable, high-performance language runtimes, a customer should have the ability to choose which components are required and the components must be decoupled one from another as much as possible.

dnakamura commented:
@samolisov  Heads up. This is probably going to break with #2530

rwy0717 commented:
Sorry, when I say a "generic test", I mean one that could work with any OMR application. Unfortunately, `gctest` really does rely on the example. You can see this in the code, eg [here](https://github.com/eclipse/omr/blob/e5db96ba52ef86f2e3b550f7b0376075602f0c99/fvtest/gctest/gcTestHelpers.hpp#L43).

By deleting the `assert(TEST OMR_EXAMPLE)`, we are no longer crashing the build in an informative way. I think we still need some kind of check:

```
if (OMR_FVTEST AND OMR_GC)
  assert(TEST OMR_EXAMPLE MESSAGE "The GC test requires the example glue, sadness")
endif()
```

Ultimately, we need to decide if we are writing tests for a specific glue/application, or tests that can work in any application (what I'm calling "generic" testing). Right now, we are only doing the former. If we are not going to make gctest "generic", maybe we should move it to the example?

rwy0717 commented on fvtest/CMakeLists.txt:
```diff
@@ -23,16 +23,10 @@ include(OmrAssert)
-omr_assert(
-	TEST OMR_EXAMPLE
-	MESSAGE "The fv test component relies on the example glue"
-)
```
This is the assert I'm concerned with removing. I don't think anyone has really thought about which tests need the example, and which don't.  gctest for sure needs the example, I suspect also rastest and vmtest. Maybe even sigtest.

samolisov replied:
I've checked a bit: if there is a custom glue library, gctest can be built without examples, so I've removed the assert from fvtest (but I had an idea to move the assert to gctest's CMakeLists.txt directly).

rwy0717 replied:
Moving the assert to gctest makes sense. The gctest definitely relies on the example's GC glue code.

charliegracie commented:
@genie-omr build all

samolisov commented:
I see many conflicts after #2533, working on these.

samolisov commented:
The branch has no conflicts with the master one anymore. As @rwy0717 said, I moved the assert to `gctest`. Also since some work was already done in #2533, I've changed some commit messages to better describe the commit reason and what happens there.

samolisov commented:
@dnakamura @charliegracie @rwy0717 could you have a look at? I've aligned the code with the solution from #2533 by @dnakamura. Ready for review.

dnakamura commented:
LGTM

charliegracie commented:
@genie-omr build all

samolisov commented:
Colleagues, all tests are passed, can you merge the PR, please? I think it would be helpful for anyone who is doing experiments with JitBuilder and the compiler and isn't working with GC.

rwy0717 commented:
LGTM.
